### PR TITLE
GP of Residual

### PR DESCRIPTION
**Pull request by jacobpkrell**
@adowling2 This PR includes all previous development but currently develops Pyomo solutions using the two-state benchmark model with a GP model of the residual correcting predictions.

### CODE REVIEW

File: examples/pyomo_tclab/data/solution_averageDraws_0o5du1.csv
Comment (jacobpkrell) [inline examples/pyomo_tclab/data/solution_averageDraws_0o5du1.csv:1]: @adowling2 Solution using average of 1000 GP draws (i.e., single Pyomo scenario where coefficients are average of GP draws). Q1 had smoothing in objective

File: examples/pyomo_tclab/data/solution_benchmark_0o5du1.csv
Comment (jacobpkrell) [inline examples/pyomo_tclab/data/solution_benchmark_0o5du1.csv:1]: @adowling2 Solution from benchmark using linear regression params from 'steps' training data. Q1 had smoothing in objective.

See '0du1' file for no smoothing.

**Comment by adowling2** [comment]
@jakekrell Can you explain the naming convention here?

**Comment by adowling2** [comment]
Results for ```solution_averageDraws_0o5du1.csv```

image

**Comment by adowling2** [comment]
@jakekrell In the notebook `tclab_benchmark.ipynb`, the objective is:

```
m.du1 = dae.DerivativeVar(m.u1)
@m.Integral(m.t)  # penalty term meant to smooth m.u1
def sum_du1(m, t):
    return m.du1[t] ** 2

@m.Objective(sense=pyo.minimize)  # objective function
def objective(m):
    return m.ise + 0.5 * m.sum_du1
```

In the notebook `tclab_optimization.ipynb`, the objective is:

```
CASE 1 - MINIMZE AVERAGE ERROR ACROSS ALL SCENARIOS:
if CASE == 1:

    @m.Objective(sense=pyo.minimize)  # objective function
    def objective(m):
        return sum(m.ise[s] for s in m.s) / len(m.s) + 0.5 * m.sum_du1

CASE 2 - MINIMIZE WORST CASE SCENARIO:
elif CASE == 2:

    m.z = pyo.Var(domain=pyo.NonNegativeReals)  # dummy variable to minimize in objective

    @m.Constraint(m.s)
    def error_threshold(m, s):
        return m.ise[s] <= m.z

    @m.Objective(sense=pyo.minimize)
    def objective(m):
        return m.z + 0.05 * m.sum_du1
```

I think we want to be careful about keeping the smoothing weight the same across the notebooks.

**Comment by adowling2** [comment]
@jakekrell I think I see why the GP is behaving strangely:
image

It is choosing an "optimal" solution that heats relies too much on the thermal mass of the heater. Notice how the heater (Th1) temperature drops off much faster than the sensor.

New modeling idea:
- Calculate the integral deviation of the sensor and temperature states from the setpoint
- Apply the GP correction to both
- Multiply the heater integral by 0.1 in the objective to give it less weight.
- This will keep the optimizer from choosing optimal profiles that cause the heater and sensor temperature to differ too much.
- Reoptimize the benchmark case using the same objective.

**Comment by adowling2** [comment]
Here are the validation experiment results using the benchmark model. 

image

**Comment by adowling2** [comment]
One more thing. Can we change the reference profile to:

```
tvec = [0, 50, 150, 450, 550, 1000]

def r(t):  # setpoint/reference
    return np.interp(t, tvec, [0, 0, 30, 30, 10, 10])
```

I suspect with Tamb + 35 we are saturating the heater.

**Comment by adowling2** [comment]
Actually, from looking at the data again, Tamb + 35 is fine. Please disregard my suggestion from a minute ago.

**Comment by jacobpkrell** [comment]
> @jakekrell Can you explain the naming convention here?

Convention is 0o5du1 for ```0.5 * m.sum_du1``` smoothing weight in objective. 0o005 means ```0.005 * m.sum_du1```, etc. Please see comment above for averageDraws meaning

**Comment by jacobpkrell** [comment]
> @jakekrell I think I see why the GP is behaving strangely: image
> 
> It is choosing an "optimal" solution that heats relies too much on the thermal mass of the heater. Notice how the heater (Th1) temperature drops off much faster than the sensor.
> 
> New modeling idea:
> 
> * Calculate the integral deviation of the sensor and temperature states from the setpoint
> * Apply the GP correction to both
> * Multiply the heater integral by 0.1 in the objective to give it less weight.
> * This will keep the optimizer from choosing optimal profiles that cause the heater and sensor temperature to differ too much.
> * Reoptimize the benchmark case using the same objective.

I like this thought. Alternatively I was exploring how to keep the GP from "blowing up" to an unreasonably large residual prediction when Th1 and Ts1 differ by a lot, but implementing this in Pyomo seems more straight forward

**Comment by adowling2** [comment]
@jakekrell Another simple idea in Pyomo is to bound R(t) between -5 and 5. The interpretation is we want to stay away from situations where our physics-based model becomes unreliable.

**Pull request merged**